File: src/main/scala/cn/academy/medicine/ItemPowder.scala
package cn.academy.medicine

import cn.academy.medicine.MatExtraction.ItemMeta
import cn.academy.medicine.Properties.Property

import java.util
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.item.{Item, ItemStack}
import net.minecraftforge.fml.common.registry.GameRegistry


//object ItemPowder {
//
//  def getProperty(stack: ItemStack): Property = stack.getItem match {
//    case item: ItemPowder => item.prop
//    case _ => throw new IllegalArgumentException("Given itemStack is not a powder")
//  }
//
//  def _init() = {
//    MatExtraction.allCombinations.foreach { case (src, prop) =>
//      val item = new ItemPowder(src, prop)
//      GameRegistry.registerItem(item, item.internalID)
//    }
//  }
//
//}

class ItemPowder(val source: ItemMeta, val prop: Property) extends Item {

//  val dummyStack = new ItemStack(source.item, 1, source.meta)

//  setTextureName("academy:powder/" + internalID)

//  def internalID = source.id.dropWhile(_ != ':').drop(1) + "_" + prop.internalID

//  override def getItemStackDisplayName(stack: ItemStack): String = {
//    source.item.getItemStackDisplayName(dummyStack) + " " + super.getItemStackDisplayName(stack)
//  }

//  override def addInformation(stack: ItemStack, player: EntityPlayer, list2: util.List[_], wtf: Boolean): Unit = {
//    val list = list2.asInstanceOf[util.List[String]]
//    list.add(prop.stackDisplayHint)
//  }
}

File: src/main/java/cn/academy/medicine/items/ItemPowder.java
package cn.academy.medicine.items;

import cn.academy.core.item.ACItem;
import cn.academy.medicine.MatExtraction;
import cn.academy.medicine.Properties;
import cpw.mods.fml.common.registry.GameRegistry;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;

import java.util.Iterator;
import java.util.List;
import java.util.Map;

public class ItemPowder extends ACItem {
    public MatExtraction.ItemMeta source;
    public Properties.Property prop;

    final ItemStack dummyStack;

    public ItemPowder(MatExtraction.ItemMeta source, Properties.Property prop) {
        super("powder");
        this.source=source;
        this.prop=prop;
        dummyStack = new ItemStack(source.item, 1, source.meta);
        setTextureName("academy:powder/" + internalID());
    }

    public String internalID()
    {
        String id = source.id;
        int test=id.indexOf(":");
        return id.substring(test+1) + "_" + prop.internalID();
    }

    @Override
    public String getItemStackDisplayName(ItemStack stack){
        return source.item.getItemStackDisplayName(dummyStack) + " " + super.getItemStackDisplayName(stack);
    }

    @Override
    public void addInformation(ItemStack stack, EntityPlayer player, List list2, boolean wtf){
        List<String> list = list2;
        list.add(prop.stackDisplayHint());
    }

    public static Properties.Property getProperty(ItemStack stack) {
        Item item = stack.getItem();
        if(item instanceof ItemPowder)
        {
            return ((ItemPowder) item).prop;
        }
        throw new IllegalArgumentException("Given itemStack is not a powder");
    }

    public static void _init(){
        for(Map.Entry<MatExtraction.ItemMeta, List<Properties.Property>> entry: MatExtraction.allRecipes.entrySet())
        {
            for(Properties.Property p:entry.getValue()) {
                ItemPowder item = new ItemPowder(entry.getKey(), p);
                GameRegistry.registerItem(item, item.internalID());
            }
        }
    }
}
